### PR TITLE
Allow specifying buffer limits in terms of byte size

### PR DESCRIPTION
**Pull request by dasch**
Since buffer size limits are mainly in place in order to avoid unbounded memory growth, defining the limits in terms of memory size makes sense. However, this does introduce some complexity – this is an attempt to see if it's worth the trouble.

### CODE REVIEW

File: lib/kafka/producer.rb [hunk 1]
```diff
@@ -218,6 +223,10 @@ def buffer_size
       @pending_messages.size + @buffer.size
     end
 
+    def buffer_bytesize
+      @pending_messages.map(&:bytesize).inject(0, &:+) + @buffer.bytesize
+    end
```
Comment (dasch) [inline lib/kafka/producer.rb]: This would need to be optimized before merging.

Comment (tcrayford) [inline lib/kafka/producer.rb]: by tracking it in a new instance var upon message insertion?

**Comment by tcrayford** [comment]
Docs work for me.

File: lib/kafka/producer.rb (hunk 1 above)
Comment (dasch) [inline lib/kafka/producer.rb]: Yup. It's just slightly messy...

Comment (dasch) [inline lib/kafka/producer.rb]: Added a smarter buffer for pending messages.

File: lib/kafka/producer.rb [hunk 2]
```diff
@@ -247,9 +264,7 @@ def shutdown
     def deliver_messages_with_retries(notification)
       attempt = 0
 
-      # Make sure we get metadata for this topic.
-      target_topics = @pending_messages.map(&:topic).uniq
-      @cluster.add_target_topics(target_topics)
+      @cluster.add_target_topics(@target_topics)
```
Comment (dasch) [inline lib/kafka/producer.rb]: Yay, killed an O(_n_) operation!

File: lib/kafka/message_buffer.rb [hunk 3]
```diff
@@ -6,22 +6,28 @@ module Kafka
...
+      @size += 1
+      @bytesize += message.bytesize
     end
 
     def concat(messages, topic:, partition:)
-      @size += messages.count
       buffer_for(topic, partition).concat(messages)
+
+      @size += messages.count
+      @bytesize += messages.map(&:bytesize).inject(0, &:+)
```
Comment (bquorning) [inline lib/kafka/message_buffer.rb]: *Slightly* faster without the `&`:
```rb
@bytesize += messages.map(&:bytesize).reduce(:+)
```

File: lib/kafka/message_buffer.rb [hunk 4]
```diff
@@ -48,6 +54,7 @@ def each
     # @return [nil]
     def clear_messages(topic:, partition:)
       @size -= @buffer[topic][partition].count
+      @bytesize -= @buffer[topic][partition].map(&:bytesize).inject(0, &:+)
```
Comment (bquorning) [inline lib/kafka/message_buffer.rb]: Same here, use `:+`.

File: lib/kafka/message_buffer.rb (hunk 3 above)
Comment (dasch) [inline lib/kafka/message_buffer.rb]: Thx.

File: lib/kafka/pending_message.rb [hunk 5]
```diff
@@ -9,5 +9,9 @@ def initialize(value:, key:, topic:, partition:, partition_key:)
       @partition = partition
       @partition_key = partition_key
     end
+
+    def bytesize
+      @key.to_s.bytesize + @value.to_s.bytesize
```
Comment (bquorning) [inline lib/kafka/pending_message.rb]: I don't understand this line.

File: lib/kafka/producer.rb [hunk 6]
```diff
@@ -178,11 +188,11 @@ def initialize(cluster:, logger:, compression_codec: nil, ack_timeout: 5, requir
     # @raise [BufferOverflow] if the maximum buffer size has been reached.
     # @return [nil]
     def produce(value, key: nil, topic:, partition: nil, partition_key: nil)
-      unless buffer_size < @max_buffer_size
+      unless buffer_size < @max_buffer_size && buffer_bytesize < @max_buffer_bytesize
         raise BufferOverflow, "Max buffer size #{@max_buffer_size} exceeded"
       end
```
Comment (bquorning) [inline lib/kafka/producer.rb]: I think the exception message needs to be changed as well.

File: lib/kafka/pending_message.rb (hunk 5 above)
Comment (dasch) [inline lib/kafka/pending_message.rb]: It's the sum of the byte size of the message's key and value. Both can be nil.

Comment (dasch) [inline lib/kafka/pending_message.rb]: I could compute the size in `initialize` and provide a reader...

File: lib/kafka/producer.rb (hunk 6 above)
Comment (dasch) [inline lib/kafka/producer.rb]: I should probably split that into two separate checks...

**Comment by codella** [comment]
:+1:

File: lib/kafka/producer.rb [hunk 7]
```diff
@@ -270,7 +285,7 @@ def deliver_messages_with_retries(notification)
         assign_partitions!
         operation.execute
 
-        if @pending_messages.empty? && @buffer.empty?
+        if @pending_message_buffer.empty? && @buffer.empty?
           break
         elsif attempt <= @max_retries
           @logger.warn "Failed to send all messages; attempting retry #{attempt} of #{@max_retries} after #{@retry_backoff}s"
```
Comment (bquorning) [inline lib/kafka/producer.rb]: Maybe `buffer_size.zero?`?

Comment (dasch) [inline lib/kafka/producer.rb]: :+1:

File: lib/kafka/producer.rb [hunk 8]
```diff
@@ -218,6 +224,17 @@ def produce(value, key: nil, topic:, partition: nil, partition_key: nil)
         partition_key: partition_key,
       )
 
+      if buffer_size >= @max_buffer_size
+        raise BufferOverflow, "Max buffer size (#{@max_buffer_size} messages) exceeded"
+      end
+
+      if buffer_bytesize + message.bytesize >= @max_buffer_bytesize
```
Comment (bquorning) [inline lib/kafka/producer.rb]: For one of the checks, you add the current `message` and not for the other. Off by one-error?

Comment (dasch) [inline lib/kafka/producer.rb]: For the count limit, we know whether we'll exceed the limit if we're currently _at it_. For the byte size limit we need to see if adding the message would put us _over the limit_.

**Pull request merged**